rl4co/envs/routing/cvrptw/generator.py

```diff
@@ -0,0 +1,159 @@
+        scale: bool = False,
+        **kwargs,
+    ):
+        super().__init__(
+            num_loc=num_loc,
+            min_loc=min_loc,
+            max_loc=max_loc,
+            loc_distribution=loc_distribution,
+            depot_distribution=depot_distribution,
+            min_demand=min_demand,
+            max_demand=max_demand,
+            demand_distribution=demand_distribution,
+            capacity=capacity,
+            **kwargs,
+        )
+
+        self.max_loc = max_loc
+        self.min_time = 0.0
+        self.max_time = max_time
+        self.scale = scale
+
+    def _generate(self, batch_size) -> TensorDict:
+        td = super()._generate(batch_size)
+
+        batch_size = [batch_size] if isinstance(batch_size, int) else batch_size
+
+        ## define service durations
+        # generate randomly (first assume service durations of 0, to be changed later)
+        durations = torch.zeros(
+            *batch_size, self.num_loc + 1, dtype=torch.float32
+        )
+
+        ## define time windows
+        # 1. get distances from depot
+        dist = get_distance(td["depot"], td["locs"].transpose(0, 1)).transpose(0, 1)
+
+        # 2. define upper bound for time windows to make sure the vehicle can get back to the depot in time
+        upper_bound = self.max_time - dist - durations
+
+        # 3. create random values between 0 and 1
+        ts_1 = torch.rand(*batch_size, self.num_loc + 1)
+        ts_2 = torch.rand(*batch_size, self.num_loc + 1)
+
+        # 4. scale values to lie between their respective min_time and max_time and convert to integer values
+        min_ts = (dist + (upper_bound - dist) * ts_1).int()
+        max_ts = (dist + (upper_bound - dist) * ts_2).int()
+
+        # 5. set the lower value to min, the higher to max
+        min_times = torch.min(min_ts, max_ts)
+        max_times = torch.max(min_ts, max_ts)
+
+        # 6. reset times for depot
+        min_times[..., :, 0] = 0.0
+        max_times[..., :, 0] = self.max_time
+
+        # 7. ensure min_times < max_times to prevent numerical errors in attention.py
+        # min_times == max_times may lead to nan values in _inner_mha()
+        mask = min_times == max_times
+        if torch.any(mask):
+            min_tmp = min_times.clone()
+            min_tmp[mask] = torch.max(
+                dist[mask].int(), min_tmp[mask] - 1
+            )  # we are handling integer values, so we can simply substract 1
+            min_times = min_tmp
+
+            mask = min_times == max_times  # update mask to new min_times
+            if torch.any(mask):
+                max_tmp = max_times.clone()
+                max_tmp[mask] = torch.min(
+                    torch.floor(upper_bound[mask]).int(),
+                    torch.max(
+                        torch.ceil(min_tmp[mask] + durations[mask]).int(),
+                        max_tmp[mask] + 1,
+                    ),
+                )
+                max_times = max_tmp
```

Comment:
We could rewrite the time window generation soonish, we have other (better) implementations in other branches/repositories already, but no hurry at all :)